CNA long format (RFC 65)

BasLee opened the pull request:
Add long data format for CNA as described in [RFC 65](https://docs.google.com/document/d/1Uk8uqXw77XZpz_1-Bi7QIHKBBfuU7aNBZ8qDfFaw3QM/edit).

Looking for feedback on the new importer and functionality that should be shared between the new ImportCnaDiscreteLongData and ImportTabDelimData.

![image](https://user-images.githubusercontent.com/3323006/195362117-17cd32e0-041f-4a9d-a9b1-a91a6f388629.png)
_Sample of new data format_

In the current CNA data format the events of each gene x sample are stored in a tabular format, with a column for each sample and a row for each gene. 

The new CNA 'long' data format contains a row for each gene x sample.

Thie new data format will import data using a new importer, the old data format is kept as is. Shared functionality should be extracted as much as possible.
  
## Changes
- Add new importer `ImportCnaDiscreteLongData` for CNA long format
- Extract shared functionality of `ImportTabDelimData` and `ImportCnaDiscreteLongData` into `CnaUtil`

![image](https://user-images.githubusercontent.com/3323006/195361924-6efd685c-2ad3-4d0e-895e-c7e3032c7d87.png)
_Database schema change_

## Tests
- Added new tests for `ImportCnaDiscreteLongData` in `TestImportCnaDiscreteLongData`

Luke-Sikina reviewed (approved):
Looks pretty good. Lots of small changes :)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/util/CnaUtil.java:
```diff
@@ -0,0 +1,108 @@
+        String entrezAsString = TabDelimitedFileUtil.getPartString(getColumnIndex(CnaUtil.ENTREZ_GENE_ID), parts);
+        if (entrezAsString.isEmpty()) {
+            return 0;
+        } else if (!entrezAsString.matches("[0-9]+")) {
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/util/CnaUtil.java:
```diff
@@ -0,0 +1,108 @@
+        List<CnaEvent> cnaEventsToAdd
+    ) throws DaoException {
+        for (CnaEvent cnaEvent : cnaEventsToAdd) {
+            if (!cnaEvent.getAlteration().equals(CNA.AMP) && !cnaEvent.getAlteration().equals(CNA.HOMDEL)) {
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTabDelimData.java:
```diff
@@ -666,24 +667,16 @@ else if (!entrez.matches("[0-9]+")) {
                                 for (CanonicalGene gene : genes) {
-                                    boolean result = storeGeneticAlterations(values, daoGeneticAlteration, gene, geneSymbol);
+                                    boolean result = this.geneticAlterationImporter.store(values, gene, geneSymbol);
                                     if (result == true) {
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTabDelimData.java:
```diff
@@ -616,7 +616,7 @@ else if (!entrez.matches("[0-9]+")) {
                         for (CanonicalGene gene : microRNAGenes) {
-                            boolean result = storeGeneticAlterations(values, daoGeneticAlteration, gene, geneSymbol);
+                            boolean result = this.geneticAlterationImporter.store(values, gene, geneSymbol);
                             if (result == true) {
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+    }
+    
+    private final int geneticProfileId;
+    private final HashSet<Long> importSetOfGenes = new HashSet<>();
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

Maybe nice to collect these kinds of preferences in style rules using [checkstyle](https://github.com/checkstyle/checkstyle)?

Luke-Sikina replied:
We already have Sonar - we could consider adding more rules to that.

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+        CanonicalGene gene,
+        String geneSymbol
+    ) throws DaoException {
+        ProfileAndGeneKey toImport = new ProfileAndGeneKey(geneticProfileId, gene.getEntrezGeneId());
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+        } catch (DaoException e) {
+            e.printStackTrace();
+        }
+    }
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

Some of the larger importers use many different dao's: injecting them all via the constructor would look quite ugly. If we want to solve this problem in all importers, maybe we want to introduce some kind of importer builders?

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+    }
+    
+    private final int geneticProfileId;
+    private final HashSet<Long> importSetOfGenes = new HashSet<>();
```
Fixed in [b9d3055](https://github.com/cBioPortal/cbioportal/pull/9847/commits/b9d3055e4e148ab2b41aadbefcba15b94ca1aa4f)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+                String geneSymbolMessage = "";
+                if (geneSymbol != null 
+                    && !geneSymbol.equalsIgnoreCase(gene.getHugoGeneSymbolAllCaps())
+                ) {
```
a special `equalsIgnoreCase` method is used here, so we cannot just swap nullable and nullsafe

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+            .row(entrezId)
+            .values()
+            .stream()
+            .filter(v -> v.geneticEvent != null)
```
Wouldn't that check if `v` (instead of v.geneticEvent`) is null?

Luke-Sikina replied:
You're right - I read this too quickly; my bad.

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+
+        long entrezId = gene.getEntrezGeneId();
+        int sampleId = sample.getInternalId();
+        CnaEventImportData eventContainer = new CnaEventImportData();
+        Table<Long, Integer, CnaEventImportData> geneBySampleEventTable = importContainer.eventsTable;
+
+        if (!geneBySampleEventTable.contains(entrezId, sample.getInternalId())) {
+            geneBySampleEventTable.put(entrezId, sampleId, eventContainer);
+        } else {
+            ProgressMonitor.logWarning(format("Skipping line %d with duplicate gene %d and sample %d", lineIndex, entrezId, sampleId));
+        }
+
+        eventContainer.geneticEvent = cnaUtil.createEvent(geneticProfile, sample.getInternalId(), lineParts);
```
Because of the new long format we only have one cna event per row (instead of all events of a single gene on one row). This would result in many more db queries, so @pvannierop and I decided to 1) collect all events, and 2) store the events in a single query . 

See also: `storeGeneticAlterations` and `storeCnaEvents`.

I'll also add some comments to explain this in the code.

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+        if (!hasData) {
+            return;
+        }
+        String[] lineParts = line.split("\t", -1);
```
Yes, maybe it would be better to use some library for this that deals with all the possible edge cases?

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+
+}
+
+class SampleIdGeneticProfileId {
```
FIxed in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+            .stream()
+            .filter(v -> v.geneticEvent != null)
+            .map(v -> v.geneticEvent)
+            .collect(toList());
```
Fixed in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+            this.extractDataToImport(geneticProfile, line, lineIndex, toImport);
+        }
+
+        DaoGeneticProfileSamples.addGeneticProfileSamples(
```
True...

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+            && geneticProfile.showProfileInAnalysisTab();
+
+        if (isDiscretizedCnaProfile) {
+            for (CnaEvent.Event event : DaoCnaEvent.getAllCnaEvents()) {
+                existingCnaEvents.put(event, event);
+            }
```
Fixed in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,340 @@
+        File cnaFile,
+        int geneticProfileId,
+        String genePanel
+    ) throws DaoException {
```
Fixed in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+                String geneSymbolMessage = "";
+                if (geneSymbol != null 
+                    && !geneSymbol.equalsIgnoreCase(gene.getHugoGeneSymbolAllCaps())
+                ) {
```
Fixed in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+        try {
+            if (!importSetOfGenes.contains(gene.getEntrezGeneId())) {
+                daoGeneticAlteration.addGeneticAlterations(geneticProfileId, gene.getEntrezGeneId(), values);
+                importSetOfGenes.add(gene.getEntrezGeneId());
```
Applied in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/GeneticAlterationImporter.java:
```diff
@@ -0,0 +1,88 @@
+
+public class GeneticAlterationImporter {
+
+    public static class ProfileAndGeneKey {
```
Fixed in [2ef2c99](https://github.com/cBioPortal/cbioportal/pull/9847/commits/2ef2c99abaf8618b1ba47e97e9c3240e8b06993e)

dippindots commented on docs/Testing.md:
````diff
@@ -92,15 +93,15 @@ SET SESSION sql_mode = 'ANSI_QUOTES';
 
 ```
-mvn integration-test -Ddb.test.username=cbio -Ddb.test.password=<your_db_password>
+mvn integration-test -Ddb.test.username=cbio_user -Ddb.test.password=somepassword
````
Great!

dippindots reviewed (commented):
Thanks @BasLee Just a few minor comments about this pr, it looks good to me.

dippindots commented:
There is one more suggestion about this pull request, since this pr introduces a new data format, we should also add a new method to validate this new data format in [validateData.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/validateData.py)

Luke-Sikina commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,333 @@
+    private String genePanel;
+    private final DaoGeneOptimized daoGene;
+    private CnaUtil cnaUtil;
+    private Map<CnaEvent.Event, CnaEvent.Event> existingCnaEvents = new HashMap<>();
```
I thought the same thing - it's because he wants to retrieve the existing cna event from the map when he finds an equivalent one, and replace the reference to the equivalent one with the original.

BasLee replied:
I reused old functionality: after a closer look, I think we can replace the map with a set. 

(The only real difference is that we now need to find the stored event in the Set, but this works fine: both a hashmap and a hashset should compare by event.hashcode and event.equals, which compare by gene and alteration.)

See [884fca6](https://github.com/cBioPortal/cbioportal/pull/9847/commits/884fca60ab44ac04dbeb27800587c0e354e54f0f).

All tests are still working.

BasLee commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,333 @@
+
+        String hugoSymbol = util.getHugoSymbol(parts);
+
+        if (Strings.isNullOrEmpty(hugoSymbol) && entrez == 0) {
```
I think it should be an `&&`; updated warning in [43f24d5](https://github.com/cBioPortal/cbioportal/pull/9847/commits/43f24d5fa933f57749f4b04361cfbd1612b55299)

BasLee commented:
> There is one more suggestion about this pull request, since this pr introduces a new data format, we should also add a new method to validate this new data format in [validateData.py](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/scripts/importer/validateData.py)

Good point! The validator should be updated. I discussed it with @pvannierop @oplantalech and we would like to do this in a separate PR.

averyniceday reviewed (commented):
overall looks good to me

are we restricting studies to have only one or the other? meaning studies can't have both a long form and standard form CNA file. This will affect whether/how we add support for recognizing the new long form CNA.

pvannierop commented:
@averyniceday Yes, we restrict a study to have only one format. I do not see how the recognition of the new format will differ from any other data types. The meta file for discrete CNA will reference the format used (long or legacy wide). Or am I overlooking a complexity here?

averyniceday commented:
@pvannierop This might be specific to how we import on our end and how the importer is implemented (in conjunction with how we organize datatypes in google sheets). We currently have a one to one mapping of meta to data files (e.g. meta_cna to data_cna). We also have a step inside the importer that merges records for genes (I imagine this is implemented to be specific to the legacy format). 

If we are not planning on introducing a new meta/data pairing for the long format (e.g. data_cna_long/meta_cna_long)  and instead only differentiating by a field inside the metafile, the importer will need to be updated to be able to differentiate between the two based on the metafile contents.

pvannierop commented:
@averyniceday The current wide format is still supported by this change (there is 100% backward compatibility). Therefore, I fail to see how this PR could possible interfere with your current processes. This PR merely adds a format. If you internally decide not to use this long format for the time being you should be fine.

And on a note of process, I think it is not really proper procedure to have merging of PRs to cBioPortal codebase to depend on update of internal tooling.

averyniceday commented:
@pvannierop @BasLee Agreed, this doesn't need to hold up merging into the cbioportal codebase. 👍 

Could you also add documentation for the new format in the File Formats section?

pvannierop commented:
@averyniceday The documentation will be added for sure! We will do this in the PR that updates the python validator script.

Luke-Sikina commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportCnaDiscreteLongData.java:
```diff
@@ -0,0 +1,331 @@
+        }
+
+        ProgressMonitor.setCurrentMessage(" --> total number of samples skipped (normal samples): " + getSamplesSkipped()
+        );
```
);

Luke-Sikina replied:
I don't really want to wait for tests again, so just fix this in the next PR. I don't really think the `);` deserves its own line.

BasLee commented:
> @BasLee CNA long format validator merged, please feel free to rebase this pr.

Rebased and fixed some validation issues, most (relevant?) checks seem to run now

BasLee commented:
After some discussion with @pvannierop: 

The new `DISCRETE_LONG` format should probably not be passed to the front end, because the old 'wide' and the new 'long' CNA format should both result in exactly the same `DISCRETE` data after the import. 

Only the importer should have knowledge of the `DISCRETE_LONG` format, and the importer should update it into `DISCRETE`. All other CNA logic can just use the old `DISCRETE` format.

These changes can be  found in  https://github.com/cBioPortal/cbioportal/pull/9847/commits/9c925246154d91e59103becab727746088fe978a
